torch/distributed/rpc.py

```diff
@@ -0,0 +1,127 @@
+    all_gather(len_outputs, len_input, group=group)
+
+    # collect name value
+    max_len = torch.stack(len_outputs).max().item()
+    name_input = torch.empty(max_len, dtype=torch.uint8)
+    name_input[:name_len] = torch.Tensor(list(name.encode('utf-8')))
+    name_outputs = [torch.empty(max_len, dtype=torch.uint8) for _ in range(ws)]
+    all_gather(name_outputs, name_input, group=group)
+
+    names = []
+    for i in range(ws):
+        name_tensor = name_outputs[i][:len_outputs[i]]
+        names.append(array.array('B', name_tensor.tolist()).tobytes().decode('utf-8'))
+
+    return names
+
+def init_rpc(name, backend='pg'):
+    r"""
+    Initialize the local RPC agent which immediately becomes ready to make and
+    accept RPCs after this method. The caller needs to make sure the specified
+    backend is properly intialized before calling this method. For example, to
+    use ``pg`` (ProcessGroup) backend, ``init_process_group`` must be invoked
+    prior to this method.
+
+    Arguments:
+        name (str): a globally unique name of the local RPC agent. It is
+                    encouraged to use names that conform application context.
+                    (e.g., ``Trainer3``, ``ParameterServer2``, ``Master``,
+                    ``Worker1``, etc.)
+        backend (str): type of RPC backend implementation. Currently,
+                       process group backend ``"pg"`` is the only available
+                       backend implementation. (default: ``"pg"``).
+    """
+    global _agent
+    if backend == 'pg':
+        from . import is_initialized
+
+        if not is_initialized():
+            raise RuntimeError("Using pg RPC backend requires calling "
+                               "init_process_group first.")
+
+        from .distributed_c10d import _default_pg
+        # TODO: issue #23232
+        names = _collect_worker_names(name, _default_pg)
+        name_dict = {names[r] : r for r in range(len(names))}
+        _agent = ProcessGroupAgent(name, name_dict, _default_pg)
+    else:
+        raise RuntimeError("Unrecognized RPC backend ", backend)
+
+
+def rpc_async(to, func, args=(), kwargs={}):
+    r"""
+    Asynchronous RPC. Make an RPC call to run function ``func`` on worker
+    ``to``, and immediately returns a future object of the return value.
+
+    Arguments:
+        to (str): name of the destination worker.
+        func (callable): a builtin function or user defined python function (e.g., ``torch.add``).
+        args (tuple): the argument tuple for the ``func`` invocation.
+        kwargs (dict): is a dictionary of keyword arguments for the ``func``
+                       invocation.
+
+    Returns:
+        A Future object that can be waited on. When completed, the return value
+        of ``func`` on ``args`` and ``kwargs`` can be retrieved from the Future
+        object.
+    """
+    qualified_name = torch.jit._find_builtin(func)
```

Comment:
we'd better assert here.
```Python
assert _agent is not None, "init_rpc(..) has not been called to setup rpc_agent yet."
```
Otherwise, the error message could look like
```Bash
Traceback (most recent call last):
  File "/usr/local/fbcode/platform007/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/local/fbcode/platform007/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/data/users/shihaoxu/fbsource/fbcode/buck-out/dev/gen/caffe2/torch/fb/modelparallel/prototype/pytorch/tests/test_rpc#binary,link
-tree/caffe2/torch/fb/modelparallel/prototype/pytorch/rpc.py", line 1057, in _peer_ping
    _low_level_rpc_no_result(_rpc_ping, inputs=[], worker_id=i)
  File "/data/users/shihaoxu/fbsource/fbcode/buck-out/dev/gen/caffe2/torch/fb/modelparallel/prototype/pytorch/tests/test_rpc#binary,link
-tree/caffe2/torch/fb/modelparallel/prototype/pytorch/rpc.py", line 484, in _low_level_rpc_no_result
    _call_rpc(RPCRequest(func, inputs, kwargs), worker_id)
  File "/data/users/shihaoxu/fbsource/fbcode/buck-out/dev/gen/caffe2/torch/fb/modelparallel/prototype/pytorch/tests/test_rpc#binary,link
-tree/caffe2/torch/fb/modelparallel/prototype/pytorch/rpc.py", line 472, in _call_rpc
    dst_name=WorkerCtx.instance().name_for_id(worker_id),
  File "/data/users/shihaoxu/fbsource/fbcode/buck-out/dev/gen/caffe2/torch/fb/modelparallel/prototype/pytorch/tests/test_rpc#binary,link
-tree/caffe2/torch/fb/modelparallel/prototype/pytorch/comm.py", line 131, in send_callable
    dist.rpc_async(dst_name, _run_request, obj)
  File "/data/users/shihaoxu/fbsource/fbcode/buck-out/dev/gen/caffe2/torch/fb/modelparallel/prototype/pytorch/tests/test_rpc#binary,link
-tree/torch/distributed/rpc.py", line 95, in rpc_async
    return invoke_rpc_python_udf(_agent, to, serialize(PythonUDF(func, args, kwargs)))
TypeError: invoke_rpc_python_udf(): incompatible function arguments. The following argument types are supported:
    1. (arg0: torch.distributed.RpcAgent, arg1: str, arg2: str) -> torch.distributed.FutureMessage

Invoked with: None, 'w:0', b'\x80\x03ctorch.distributed.internal_rpc_utils\nPythonUDF\nq\x00ccaffe2.torch.fb.modelparallel.prototype.pyt
orch.comm\n_run_request\nq\x01ccaffe2.torch.fb.modelparallel.prototype.pytorch.rpc\nRPCRequest\nq\x02)\x81q\x03}q\x04(X\x04\x00\x00\x00f
uncq\x05ccaffe2.torch.fb.modelparallel.prototype.pytorch.rpc\n_rpc_ping\nq\x06X\x06\x00\x00\x00callerq\x07X\x03\x00\x00\x00w:1q\x08X\x06
\x00\x00\x00outputq\tNX\x06\x00\x00\x00inputsq\n]q\x0bX\x06\x00\x00\x00kwargsq\x0cNX\x0c\x00\x00\x00client_scopeq\rccaffe2.torch.fb.mode
lparallel.prototype.pytorch.rpc\nRRefId\nq\x0eK\x01K\x00\x86q\x0f\x81q\x10ub}q\x11\x87q\x12\x81q\x13.'
Traceback (most recent call last):
```

Reply:
in which case _agent could be None? maybe we should fix that first

Reply:
in which case _agent is None? maybe we should fix that first

Reply:
in which case _agent will be None? maybe we should fix that first